services/flake_detection.py

```diff
@@ -0,0 +1,235 @@
+            ):
+                self.curr_test_context.is_curr_flake = True
+                self.resulting_flakes[
+                    self.curr_test_context.curr_test_id
+                ] = FlakeType.ConsecutiveDiffOutcomes
+
+                metrics.incr(
+                    "flake_detection.detect_flakes.flake_detected",
+                    1,
+                    tags={"flake_type": str(FlakeType.ConsecutiveDiffOutcomes)},
+                )
+                return True
+            elif existing_outcome_on_commit is None:
+                self.curr_test_context.commit_to_outcome[
+                    instance.commitid
+                ] = instance.outcome
+        return False
+
+    def check_if_failure_messages_match(self, instance):
+        failure_message = instance.failure_message
+        if self.failure_normalizer is not None:
+            failure_message = self.failure_normalizer.normalize_failure_message(
+                failure_message
+            )
+
+        self.curr_test_context.flake_dict[failure_message].counter += 1
+        self.curr_test_context.flake_dict[failure_message].branch.add(instance.branch)
+        self.curr_test_context.flake_dict[failure_message].time.append(
+            instance.created_at.timestamp()
+        )
+
+        potential_flake = self.curr_test_context.flake_dict[failure_message]
+        if potential_flake.counter > 1 and len(potential_flake.branch) > 2:
+            # Exact error happened on 2 other branches at least
+            self.curr_test_context.is_curr_flake = True
+            self.resulting_flakes[
+                self.curr_test_context.curr_test_id
+            ] = FlakeType.UnrelatedMatchingFailures
+
+            metrics.incr(
+                "flake_detection.detect_flakes.flake_detected",
+                1,
+                tags={"flake_type": str(FlakeType.UnrelatedMatchingFailures)},
+            )
+            return True
+        return False
+
+    @trace
+    def detect_flakes(self) -> List[Tuple[str, str]]:
+        """
+        Detect flaky tests on a given repo based on the test instances
+        gathered in the query in the constructor
+
+        A test is a flake if:
+        - it has failed on default branch
+        OR
+        - it has different outcomes on different runs for the same commit
+        OR
+        - it has the same failure messages on failures on unrelated branches
+
+        We check which kind of flake exists in that order
+
+
+        TODO: Differentiate between random flakes and infra flakes. Infra flakes
+        are flakes that are caused by a dependency or a temporary outage.
+        They can be defined as: matching failure messages on unrelated branches
+        clustered at a point in time, this test should only be failing for that
+        range in time.
+
+        Returns a list of tuples of flaky test id and type of flake
+        """
+        self.curr_test_context = FlakeDetectionContext()
```

Comment:
[very nit] This is a personal preference, you can ignore it completely if you don't feel the same way.

Personally I'd have made `curr_test_context` just a variable of this function and passed it around. It's easier to write tests to individual functions that use the state (easier to mock the context), although only slightly.

But when reading the code I don't have to look into where this context is defined in the class (and it's not in the `__init__` method, so it's hard to find) if it's just another arg, with it's type right there.

One last reason is because of the garbage collection. Having this a local var will allow the context to the cleaned faster when the function finishes (if the class lives longer).

You don't have to change anything, just something to think about.